<a href="https://colab.research.google.com/github/LokeRuiKee/ml-dl-playground/blob/main/GAN_experiment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# GAN Template Code

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])


In [ ]:
# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')


In [ ]:
# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)


In [ ]:
# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)



2/2 [==============================] - 0s 6ms/step
Epoch 1/100, Discriminator Loss: 0.5023490190505981, Generator Loss: 1.774107575416565
2/2 [==============================] - 0s 5ms/step
Epoch 2/100, Discriminator Loss: 0.4587470665574074, Generator Loss: 2.231952428817749
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 0.475782111287117, Generator Loss: 2.683269500732422
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.389856293797493, Generator Loss: 3.0650861263275146
2/2 [==============================] - 0s 6ms/step
Epoch 5/100, Discriminator Loss: 0.4386894963681698, Generator Loss: 3.4051804542541504
2/2 [==============================] - 0s 5ms/step
Epoch 6/100, Discriminator Loss: 0.4262126889079809, Generator Loss: 3.681575298309326
2/2 [==============================] - 0s 6ms/step
Epoch 7/100, Discriminator Loss: 0.36120539251714945, Generator Loss: 3.901822090148926
2/2 [============================

# 7 activation, optimizer = adam

output activation = sigmoid

## relu

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 10ms/step
Epoch 1/100, Discriminator Loss: 0.8483355045318604, Generator Loss: 0.4371866285800934
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.6103552430868149, Generator Loss: 0.6390867233276367
2/2 [==============================] - 0s 8ms/step
Epoch 3/100, Discriminator Loss: 0.5041152387857437, Generator Loss: 0.8884485960006714
2/2 [==============================] - 0s 7ms/step
Epoch 4/100, Discriminator Loss: 0.40610338747501373, Generator Loss: 1.1705973148345947
2/2 [==============================] - 0s 7ms/step
Epoch 5/100, Discriminator Loss: 0.3498876243829727, Generator Loss: 1.4392814636230469
2/2 [==============================] - 0s 5ms/step
Epoch 6/100, Discriminator Loss: 0.27148105204105377, Generator Loss: 1.7357869148254395
2/2 [==============================] - 0s 9ms/step
Epoch 7/100, Discriminator Loss: 0.2224211022257805, Generator Loss: 1.996477723121643
2/2 [====================

## LeakyReLU

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='LeakyReLU'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='LeakyReLU'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 6ms/step
Epoch 1/100, Discriminator Loss: 0.6666022539138794, Generator Loss: 1.1098394393920898
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.5471976846456528, Generator Loss: 1.7239251136779785
2/2 [==============================] - 0s 6ms/step
Epoch 3/100, Discriminator Loss: 0.41529730707407, Generator Loss: 2.395282506942749
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.40821099281311035, Generator Loss: 2.9738125801086426
2/2 [==============================] - 0s 6ms/step
Epoch 5/100, Discriminator Loss: 0.39681243523955345, Generator Loss: 3.49904203414917
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 0.37849222123622894, Generator Loss: 3.885695695877075
2/2 [==============================] - 0s 5ms/step
Epoch 7/100, Discriminator Loss: 0.4418186154216528, Generator Loss: 4.23897123336792
2/2 [===========================

## elu

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='elu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='elu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 6ms/step
Epoch 1/100, Discriminator Loss: 0.9392116665840149, Generator Loss: 0.6346926689147949
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.6651347875595093, Generator Loss: 1.2459287643432617
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 0.4726167395710945, Generator Loss: 1.9961163997650146
2/2 [==============================] - 0s 4ms/step
Epoch 4/100, Discriminator Loss: 0.4339911565184593, Generator Loss: 2.743772029876709
2/2 [==============================] - 0s 5ms/step
Epoch 5/100, Discriminator Loss: 0.3639991991221905, Generator Loss: 3.4275054931640625
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 0.32620599679648876, Generator Loss: 3.977827310562134
2/2 [==============================] - 0s 6ms/step
Epoch 7/100, Discriminator Loss: 0.3685890920460224, Generator Loss: 4.45877742767334
2/2 [=========================

## PReLU

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='PReLU'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='PReLU'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 5ms/step
Epoch 1/100, Discriminator Loss: 0.7371473014354706, Generator Loss: 1.2192940711975098
2/2 [==============================] - 0s 7ms/step
Epoch 2/100, Discriminator Loss: 0.6483373790979385, Generator Loss: 1.6305361986160278
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 0.6321543231606483, Generator Loss: 2.08355712890625
2/2 [==============================] - 0s 5ms/step
Epoch 4/100, Discriminator Loss: 0.5813784152269363, Generator Loss: 2.5080819129943848
2/2 [==============================] - 0s 5ms/step
Epoch 5/100, Discriminator Loss: 0.49484626576304436, Generator Loss: 2.908904552459717
2/2 [==============================] - 0s 5ms/step
Epoch 6/100, Discriminator Loss: 0.5099383257329464, Generator Loss: 3.2484307289123535
2/2 [==============================] - 0s 6ms/step
Epoch 7/100, Discriminator Loss: 0.4972410462796688, Generator Loss: 3.548252820968628
2/2 [=========================

## Sigmoid

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='sigmoid'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='sigmoid'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 6ms/step
Epoch 1/100, Discriminator Loss: 0.7191917151212692, Generator Loss: 1.1138420104980469
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.6853619664907455, Generator Loss: 1.2956457138061523
2/2 [==============================] - 0s 6ms/step
Epoch 3/100, Discriminator Loss: 0.6379567384719849, Generator Loss: 1.4926944971084595
2/2 [==============================] - 0s 4ms/step
Epoch 4/100, Discriminator Loss: 0.5993589386343956, Generator Loss: 1.6947638988494873
2/2 [==============================] - 0s 9ms/step
Epoch 5/100, Discriminator Loss: 0.5655164420604706, Generator Loss: 1.890343189239502
2/2 [==============================] - 0s 4ms/step
Epoch 6/100, Discriminator Loss: 0.564853273332119, Generator Loss: 2.074108123779297
2/2 [==============================] - 0s 6ms/step
Epoch 7/100, Discriminator Loss: 0.5680992603302002, Generator Loss: 2.246067523956299
2/2 [==========================

## tanh

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='tanh'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='tanh'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 5ms/step
Epoch 1/100, Discriminator Loss: 0.8165392279624939, Generator Loss: 0.8608134388923645
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.6211243867874146, Generator Loss: 1.481255292892456
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 0.456589512526989, Generator Loss: 2.194389820098877
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.42766691371798515, Generator Loss: 2.8565773963928223
2/2 [==============================] - 0s 7ms/step
Epoch 5/100, Discriminator Loss: 0.421856414526701, Generator Loss: 3.4424867630004883
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 0.4326293608173728, Generator Loss: 3.9297263622283936
2/2 [==============================] - 0s 6ms/step
Epoch 7/100, Discriminator Loss: 0.4069856833666563, Generator Loss: 4.333117485046387
2/2 [==========================

## softmax

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='softmax'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='softmax'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 7ms/step
Epoch 1/100, Discriminator Loss: 0.6929576694965363, Generator Loss: 0.6915794610977173
2/2 [==============================] - 0s 5ms/step
Epoch 2/100, Discriminator Loss: 0.6903967261314392, Generator Loss: 0.6962755918502808
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 0.6887102723121643, Generator Loss: 0.7010542154312134
2/2 [==============================] - 0s 4ms/step
Epoch 4/100, Discriminator Loss: 0.6864204406738281, Generator Loss: 0.705707311630249
2/2 [==============================] - 0s 5ms/step
Epoch 5/100, Discriminator Loss: 0.6848939657211304, Generator Loss: 0.7101559638977051
2/2 [==============================] - 0s 5ms/step
Epoch 6/100, Discriminator Loss: 0.6817993521690369, Generator Loss: 0.7143985033035278
2/2 [==============================] - 0s 7ms/step
Epoch 7/100, Discriminator Loss: 0.6785003244876862, Generator Loss: 0.7184659242630005
2/2 [=======================

## conclusion
Best activation function = relu.
Epoch 100/100, Discriminator Loss: 0.06684121955186129, Generator Loss: 3.91917085647583 (?)

# Best activation function: relu, 8 optimizer

## SGD

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import SGD
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=SGD(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=SGD(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 6ms/step
Epoch 1/100, Discriminator Loss: 0.48816950619220734, Generator Loss: 1.1955674886703491
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.46098193526268005, Generator Loss: 1.1976029872894287
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 0.467398539185524, Generator Loss: 1.201136827468872
2/2 [==============================] - 0s 4ms/step
Epoch 4/100, Discriminator Loss: 0.49841998517513275, Generator Loss: 1.226931095123291
2/2 [==============================] - 0s 6ms/step
Epoch 5/100, Discriminator Loss: 0.500644788146019, Generator Loss: 1.219269037246704
2/2 [==============================] - 0s 5ms/step
Epoch 6/100, Discriminator Loss: 0.4449627995491028, Generator Loss: 1.2356300354003906
2/2 [==============================] - 0s 5ms/step
Epoch 7/100, Discriminator Loss: 0.4500531852245331, Generator Loss: 1.2460296154022217
2/2 [========================

## SGD with Momentum

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import SGD
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=SGD(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=SGD(learning_rate=0.0002, momentum=0.9), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 5ms/step
Epoch 1/100, Discriminator Loss: 0.9617482125759125, Generator Loss: 0.4084407091140747
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.9347191751003265, Generator Loss: 0.41926395893096924
2/2 [==============================] - 0s 6ms/step
Epoch 3/100, Discriminator Loss: 0.9230595827102661, Generator Loss: 0.4281107783317566
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.8794839978218079, Generator Loss: 0.4376775622367859
2/2 [==============================] - 0s 7ms/step
Epoch 5/100, Discriminator Loss: 0.8949667513370514, Generator Loss: 0.43996021151542664
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 0.9068942070007324, Generator Loss: 0.4532725214958191
2/2 [==============================] - 0s 5ms/step
Epoch 7/100, Discriminator Loss: 0.8473237752914429, Generator Loss: 0.4658412039279938
2/2 [====================

## RMSProp

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import RMSprop
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=RMSprop(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=RMSprop(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 9ms/step
Epoch 1/100, Discriminator Loss: 0.6673113703727722, Generator Loss: 2.0603761672973633
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.40881116688251495, Generator Loss: 2.4071719646453857
2/2 [==============================] - 0s 7ms/step
Epoch 3/100, Discriminator Loss: 0.43188953772187233, Generator Loss: 2.6917238235473633
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.4104255586862564, Generator Loss: 2.8758630752563477
2/2 [==============================] - 0s 6ms/step
Epoch 5/100, Discriminator Loss: 0.39298442006111145, Generator Loss: 3.0295894145965576
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 0.346389502286911, Generator Loss: 3.191779613494873
2/2 [==============================] - 0s 5ms/step
Epoch 7/100, Discriminator Loss: 0.3939501829445362, Generator Loss: 3.275360584259033
2/2 [======================

## Adam

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 5ms/step
Epoch 1/100, Discriminator Loss: 1.0912537276744843, Generator Loss: 0.8375904560089111
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.9141736626625061, Generator Loss: 1.2766445875167847
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 0.8652187436819077, Generator Loss: 1.7646663188934326
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.8285566121339798, Generator Loss: 2.2488136291503906
2/2 [==============================] - 0s 7ms/step
Epoch 5/100, Discriminator Loss: 0.8103710040450096, Generator Loss: 2.7206499576568604
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 0.7277082670480013, Generator Loss: 3.1161065101623535
2/2 [==============================] - 0s 6ms/step
Epoch 7/100, Discriminator Loss: 0.7112441658973694, Generator Loss: 3.4289469718933105
2/2 [======================

## Adamax

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adamax
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adamax(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adamax(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 10ms/step
Epoch 1/100, Discriminator Loss: 0.7871482074260712, Generator Loss: 0.6588345766067505
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.6783224642276764, Generator Loss: 0.8688931465148926
2/2 [==============================] - 0s 6ms/step
Epoch 3/100, Discriminator Loss: 0.662421315908432, Generator Loss: 1.0964429378509521
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.5785346031188965, Generator Loss: 1.3306801319122314
2/2 [==============================] - 0s 5ms/step
Epoch 5/100, Discriminator Loss: 0.5217867791652679, Generator Loss: 1.5755902528762817
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 0.4863128513097763, Generator Loss: 1.7727508544921875
2/2 [==============================] - 0s 8ms/step
Epoch 7/100, Discriminator Loss: 0.38719528913497925, Generator Loss: 1.9677329063415527
2/2 [=====================

## Nadam

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Nadam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Nadam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Nadam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 6ms/step
Epoch 1/100, Discriminator Loss: 0.517255574464798, Generator Loss: 1.6962848901748657
2/2 [==============================] - 0s 5ms/step
Epoch 2/100, Discriminator Loss: 0.4842006489634514, Generator Loss: 1.9504759311676025
2/2 [==============================] - 0s 7ms/step
Epoch 3/100, Discriminator Loss: 0.45045557618141174, Generator Loss: 2.171443462371826
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.4428761526942253, Generator Loss: 2.4050846099853516
2/2 [==============================] - 0s 6ms/step
Epoch 5/100, Discriminator Loss: 0.48130281269550323, Generator Loss: 2.643372058868408
2/2 [==============================] - 0s 7ms/step
Epoch 6/100, Discriminator Loss: 0.4526028484106064, Generator Loss: 2.853687286376953
2/2 [==============================] - 0s 7ms/step
Epoch 7/100, Discriminator Loss: 0.35573171451687813, Generator Loss: 3.0525307655334473
2/2 [=======================

## Adagrad

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from keras.optimizers import Adagrad

# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adagrad(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adagrad(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 9ms/step
Epoch 1/100, Discriminator Loss: 1.1328797340393066, Generator Loss: 1.3906285762786865
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 1.1036754250526428, Generator Loss: 1.4057244062423706
2/2 [==============================] - 0s 6ms/step
Epoch 3/100, Discriminator Loss: 1.0293585807085037, Generator Loss: 1.435567855834961
2/2 [==============================] - 0s 7ms/step
Epoch 4/100, Discriminator Loss: 1.150167465209961, Generator Loss: 1.4592020511627197
2/2 [==============================] - 0s 5ms/step
Epoch 5/100, Discriminator Loss: 1.087673395872116, Generator Loss: 1.4797892570495605
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 0.9341898411512375, Generator Loss: 1.501943588256836
2/2 [==============================] - 0s 11ms/step
Epoch 7/100, Discriminator Loss: 1.0738927870988846, Generator Loss: 1.5249216556549072
2/2 [=========================

## AdaDelta

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from keras.optimizers import Adadelta

# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adadelta(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adadelta(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 5ms/step
Epoch 1/100, Discriminator Loss: 0.7126422822475433, Generator Loss: 1.1439472436904907
2/2 [==============================] - 0s 7ms/step
Epoch 2/100, Discriminator Loss: 0.7737583816051483, Generator Loss: 1.1377840042114258
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 0.7376970499753952, Generator Loss: 1.141035556793213
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.7167271971702576, Generator Loss: 1.142885684967041
2/2 [==============================] - 0s 6ms/step
Epoch 5/100, Discriminator Loss: 0.7954206764698029, Generator Loss: 1.1386210918426514
2/2 [==============================] - 0s 5ms/step
Epoch 6/100, Discriminator Loss: 0.7253158837556839, Generator Loss: 1.1329257488250732
2/2 [==============================] - 0s 5ms/step
Epoch 7/100, Discriminator Loss: 0.755130872130394, Generator Loss: 1.1369915008544922
2/2 [=========================

## conclusion
Best optimizer = adam (commonly used in GAN).
Epoch 100/100, Discriminator Loss: 0.1789083518087864, Generator Loss: 4.167590141296387

#Experimenting Best Output Activation Function


## output relu

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='relu'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='relu')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 9ms/step
Epoch 1/100, Discriminator Loss: 6.0574450343847275, Generator Loss: 7.489331245422363
2/2 [==============================] - 0s 10ms/step
Epoch 2/100, Discriminator Loss: 4.91982177644968, Generator Loss: 7.552118301391602
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 4.955460950732231, Generator Loss: 5.818508625030518
2/2 [==============================] - 0s 7ms/step
Epoch 4/100, Discriminator Loss: 5.2745649963617325, Generator Loss: 5.218440055847168
2/2 [==============================] - 0s 10ms/step
Epoch 5/100, Discriminator Loss: 4.901006415486336, Generator Loss: 4.9548115730285645
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 4.441048711538315, Generator Loss: 4.2838826179504395
2/2 [==============================] - 0s 6ms/step
Epoch 7/100, Discriminator Loss: 4.25408798456192, Generator Loss: 4.043478965759277
2/2 [==============================] 

## output LeakyReLU

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='LeakyReLU'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='LeakyReLU')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 5ms/step
Epoch 1/100, Discriminator Loss: 4.497592464089394, Generator Loss: 5.717559814453125
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 5.219203278422356, Generator Loss: 5.034558296203613
2/2 [==============================] - 0s 6ms/step
Epoch 3/100, Discriminator Loss: 4.56376188993454, Generator Loss: 3.2234902381896973
2/2 [==============================] - 0s 5ms/step
Epoch 4/100, Discriminator Loss: 5.703654408454895, Generator Loss: 1.9002270698547363
2/2 [==============================] - 0s 5ms/step
Epoch 5/100, Discriminator Loss: 4.570112228393555, Generator Loss: 0.9297850131988525
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 4.655741989612579, Generator Loss: 1.0033323764801025
2/2 [==============================] - 0s 8ms/step
Epoch 7/100, Discriminator Loss: 5.305236458778381, Generator Loss: 0.668152928352356
2/2 [==============================] -

## output elu

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='elu'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='elu')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 9ms/step
Epoch 1/100, Discriminator Loss: 6.291070997714996, Generator Loss: 9.230876922607422
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 6.703990668058395, Generator Loss: 5.426397323608398
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 6.292546480894089, Generator Loss: 4.313553810119629
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 6.054940104484558, Generator Loss: 3.719538688659668
2/2 [==============================] - 0s 6ms/step
Epoch 5/100, Discriminator Loss: 6.483374506235123, Generator Loss: 2.92628812789917
2/2 [==============================] - 0s 8ms/step
Epoch 6/100, Discriminator Loss: 6.155518084764481, Generator Loss: 1.6110055446624756
2/2 [==============================] - 0s 5ms/step
Epoch 7/100, Discriminator Loss: 6.4097461104393005, Generator Loss: 1.2838153839111328
2/2 [==============================] - 

## output PReLU

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='PReLU'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='PReLU')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 6ms/step
Epoch 1/100, Discriminator Loss: 6.071852080523968, Generator Loss: 11.948670387268066
2/2 [==============================] - 0s 5ms/step
Epoch 2/100, Discriminator Loss: 6.050487671047449, Generator Loss: 10.022926330566406
2/2 [==============================] - 0s 6ms/step
Epoch 3/100, Discriminator Loss: 6.067810297012329, Generator Loss: 7.469822883605957
2/2 [==============================] - 0s 10ms/step
Epoch 4/100, Discriminator Loss: 6.2075101509690285, Generator Loss: 8.207794189453125
2/2 [==============================] - 0s 5ms/step
Epoch 5/100, Discriminator Loss: 5.722714081406593, Generator Loss: 6.514898300170898
2/2 [==============================] - 0s 10ms/step
Epoch 6/100, Discriminator Loss: 5.7138590812683105, Generator Loss: 6.087976455688477
2/2 [==============================] - 0s 10ms/step
Epoch 7/100, Discriminator Loss: 6.385417595505714, Generator Loss: 5.163500785827637
2/2 [=============================

## output sigmoid

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='sigmoid'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 6ms/step
Epoch 1/100, Discriminator Loss: 1.1097444891929626, Generator Loss: 0.41673368215560913
2/2 [==============================] - 0s 5ms/step
Epoch 2/100, Discriminator Loss: 0.9179717898368835, Generator Loss: 0.6204590797424316
2/2 [==============================] - 0s 5ms/step
Epoch 3/100, Discriminator Loss: 0.6926451921463013, Generator Loss: 0.8950811624526978
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.5338695794343948, Generator Loss: 1.2148325443267822
2/2 [==============================] - 0s 5ms/step
Epoch 5/100, Discriminator Loss: 0.5499390363693237, Generator Loss: 1.5115671157836914
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 0.5002485364675522, Generator Loss: 1.7907359600067139
2/2 [==============================] - 0s 6ms/step
Epoch 7/100, Discriminator Loss: 0.4873367100954056, Generator Loss: 2.0511555671691895
2/2 [=====================

## output tanh

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='tanh'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='tanh')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 6ms/step
Epoch 1/100, Discriminator Loss: 5.843226844444871, Generator Loss: 13.424934387207031
2/2 [==============================] - 0s 5ms/step
Epoch 2/100, Discriminator Loss: 6.847088024020195, Generator Loss: 10.910512924194336
2/2 [==============================] - 0s 6ms/step
Epoch 3/100, Discriminator Loss: 5.639895021915436, Generator Loss: 7.778335094451904
2/2 [==============================] - 0s 5ms/step
Epoch 4/100, Discriminator Loss: 5.635799504816532, Generator Loss: 6.488965034484863
2/2 [==============================] - 0s 6ms/step
Epoch 5/100, Discriminator Loss: 5.991872280836105, Generator Loss: 6.761526107788086
2/2 [==============================] - 0s 6ms/step
Epoch 6/100, Discriminator Loss: 6.1392310708761215, Generator Loss: 4.433080673217773
2/2 [==============================] - 0s 5ms/step
Epoch 7/100, Discriminator Loss: 5.643904685974121, Generator Loss: 2.5555224418640137
2/2 [==============================] 

## output softmax

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Dense, Flatten, Reshape
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
# Define the dimensions of the latent space (input to the generator)
latent_dim = 100
# Define the generator model
generator = Sequential([
    Dense(128, input_dim=latent_dim, activation='relu'),
    Dense(784, activation='softmax'),
    Reshape((28, 28))
])
# Define the discriminator model
discriminator = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(1, activation='softmax')
])

# Compile the discriminator (binary crossentropy loss)
discriminator.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy', metrics=['accuracy'])
# Freeze the discriminator during the combined model training
discriminator.trainable = False
# Define the combined model (GAN)
gan = Sequential([generator, discriminator])
# Compile the GAN (binary crossentropy loss)
gan.compile(optimizer=Adam(learning_rate=0.0002), loss='binary_crossentropy')

# Function to generate random noise (latent points)
def generate_latent_points(latent_dim, n_samples):
    return np.random.randn(n_samples, latent_dim)
# Function to generate fake images using the generator
def generate_fake_samples(generator, latent_dim, n_samples):
    latent_points = generate_latent_points(latent_dim, n_samples)
    return generator.predict(latent_points)

# Function to train the GAN
def train_gan(gan, generator, discriminator, latent_dim, n_epochs=100, batch_size=128):
    half_batch = batch_size // 2
    for epoch in range(n_epochs):
        # Train the discriminator
        real_images = np.random.randn(half_batch, 28, 28)
        real_labels = np.ones((half_batch, 1))
        fake_images = generate_fake_samples(generator, latent_dim, half_batch)
        fake_labels = np.zeros((half_batch, 1))
        discriminator_loss_real = discriminator.train_on_batch(real_images, real_labels)
        discriminator_loss_fake = discriminator.train_on_batch(fake_images, fake_labels)
        discriminator_loss = 0.5 * np.add(discriminator_loss_real, discriminator_loss_fake)
        # Train the generator (via the combined GAN model)
        latent_points = generate_latent_points(latent_dim, batch_size)
        misleading_targets = np.ones((batch_size, 1))
        gan_loss = gan.train_on_batch(latent_points, misleading_targets)
        # Print progress
        print(f"Epoch {epoch+1}/{n_epochs}, Discriminator Loss: {discriminator_loss[0]}, Generator Loss: {gan_loss}")

# Train the GAN
train_gan(gan, generator, discriminator, latent_dim)

2/2 [==============================] - 0s 5ms/step
Epoch 1/100, Discriminator Loss: 0.6806151270866394, Generator Loss: 0.6926451921463013
2/2 [==============================] - 0s 6ms/step
Epoch 2/100, Discriminator Loss: 0.7304114103317261, Generator Loss: 0.6926403641700745
2/2 [==============================] - 0s 6ms/step
Epoch 3/100, Discriminator Loss: 0.6629149913787842, Generator Loss: 0.6927647590637207
2/2 [==============================] - 0s 6ms/step
Epoch 4/100, Discriminator Loss: 0.7015375792980194, Generator Loss: 0.692947268486023
2/2 [==============================] - 0s 5ms/step
Epoch 5/100, Discriminator Loss: 0.6804099082946777, Generator Loss: 0.6930649280548096
2/2 [==============================] - 0s 5ms/step
Epoch 6/100, Discriminator Loss: 0.7300579249858856, Generator Loss: 0.6932889223098755
2/2 [==============================] - 0s 6ms/step
Epoch 7/100, Discriminator Loss: 0.7556909620761871, Generator Loss: 0.6934186816215515
2/2 [=======================

## conclusion
Best output activation function = softmax (?). Epoch 100/100, Discriminator Loss: 0.4153352379798889, Generator Loss: 0.7145787477493286


**explaination from chatgpt. to be verified the credibility:

To determine the "best" result:

The discriminator loss should ideally be low, indicating that the discriminator can distinguish well between real and fake samples.

The generator loss should ideally be high, indicating that the generator can produce samples that are challenging for the discriminator to classify as fake.

From the provided results, the last epoch (Epoch 100/100) with the discriminator loss of 0.415 and the generator loss of 0.715 seems to be the best result.

This is because the discriminator loss is relatively low, indicating good discrimination capability, and the generator loss is moderately high, indicating that the generator is producing samples that are challenging for the discriminator.